### ISSUE TITLE
🐛 Bug Report: generate-patch doesn't work with yarn backstage plugin

### ISSUE THREAD

**Issue by tcardonne** [backstage/backstage#27940]
### 📜 Description

The `yarn backstage-repo-tools generate-patch` command doesn't correctly update the `resolutions` in the `package.json` file when applying a patch.

### 👍 Expected behavior

The command correctly recognises that the Yarn plugin is installed and adds a correct resolution so the patch is correctly applied.

### 👎 Actual Behavior with Screenshots

The resolution that is added is:

```
"resolutions": {
    "@types/react": "^18",
    "@types/react-dom": "^18",
    "@backstage/plugin-catalog@npm:^1.25.1": "patch:%40backstage%2Fplugin-catalog%40npm%3A1.25.1#.yarn/patches/@backstage-plugin-catalog-1.25.1-e89242027d65b-dirty.patch"
  },
```

but then if I do a `yarn why @backstage/plugin-catalog`:

```
❯ yarn why @backstage/plugin-catalog
├─ @backstage/plugin-api-docs@npm:0.12.1
│  └─ @backstage/plugin-catalog@patch:@backstage/plugin-catalog@npm%3A1.25.1#.yarn/patches/@backstage-plugin-catalog-1.25.1-e89242027d65b-dirty.patch::version=1.25.1&hash=26757f&locator=root%40workspace%3A. (via patch:%40backstage%2Fplugin-catalog%40npm%3A1.25.1#.yarn/patches/@backstage-plugin-catalog-1.25.1-e89242027d65b-dirty.patch::locator=root%40workspace%3A.)
│
├─ @backstage/plugin-api-docs@npm:0.12.1 [f87a9]
│  └─ @backstage/plugin-catalog@patch:@backstage/plugin-catalog@npm%3A1.25.1#.yarn/patches/@backstage-plugin-catalog-1.25.1-e89242027d65b-dirty.patch::version=1.25.1&hash=26757f&locator=root%40workspace%3A. [4afe9] (via patch:%40backstage%2Fplugin-catalog%40npm%3A1.25.1#.yarn/patches/@backstage-plugin-catalog-1.25.1-e89242027d65b-dirty.patch::locator=root%40workspace%3A. [4afe9])
│
└─ app@workspace:packages/app
   └─ @backstage/plugin-catalog@npm:1.25.1 [f87a9] (via backstage:1.33.5 [f87a9])
```

 `packages/app` uses `@npm:1.25.1` instead of the patch.

### 👟 Reproduction steps

Have `backstage/backstage` cloned and bootstrap a new Backstage instance somewhere (we'll call it `demo`).

```
npx @backstage/create-app

# in the demo instance
# make sure to use the plugin
yarn plugin import https://versions.backstage.io/v1/tags/main/yarn-plugin
yarn backstage-cli versions:bump
yarn install
```

Edit a plugin in the Backstage repository (eg, `plugins/catalog/`).

Generate the patch using `backstage-repo-tools`:

```
yarn backstage-repo-tools generate-patch --target ../../backstage/demo --base-version "1.25.1" @backstage/plugin-catalog
```

The patch is created and a resolution is added, but the patch isn't used by `yarn install`.

You can verify it using `yarn why`.

### 📃 Provide the context for the Bug.

If I correct the resolution to 

```
"resolutions": {
    "@types/react": "^18",
    "@types/react-dom": "^18",
    "@backstage/plugin-catalog@npm:^1.25.1": "patch:%40backstage%2Fplugin-catalog%40npm%3A1.25.1#.yarn/patches/@backstage-plugin-catalog-1.25.1-e89242027d65b-dirty.patch",
    "@backstage/plugin-catalog@backstage:^": "patch:%40backstage%2Fplugin-catalog%40npm%3A1.25.1#.yarn/patches/@backstage-plugin-catalog-1.25.1-e89242027d65b-dirty.patch"
  },
```

the patch is correctly applied.

I need to have both to patch the direct dependency of `packages/app` (via `backstage:^` and nested dependencies (via `@npm:^1.25.1`)

### 🖥️ Your Environment

```
❯ yarn backstage-cli --version
0.29.3-next.0

❯ yarn backstage-repo-tools --version                                                                                     
0.12.0-next.0
```

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes, I'll try to understand how it works but if someone has a hint I'll take it!

**Comment by freben** [comment]
A more brutal approach could be to skip the version selector

```json
"resolutions": {
    "@types/react": "^18",
    "@types/react-dom": "^18",
    "@backstage/plugin-catalog": "patch:%40backstage%2Fplugin-catalog%40npm%3A1.25.1#.yarn/patches/@backstage-plugin-catalog-1.25.1-e89242027d65b-dirty.patch",
    "@backstage/plugin-catalog@backstage:^": "patch:%40backstage%2Fplugin-catalog%40npm%3A1.25.1#.yarn/patches/@backstage-plugin-catalog-1.25.1-e89242027d65b-dirty.patch"
  },
```

In not sure which is better. Matching on the `backstage` selector kind of has the same drawback right?

**Comment by Rugvip** [comment]
Gonna close this one as fixed by #27952, since it no longer cares about descriptors and just adds an override for the entire package instead
